content/en/BestPractices/SSTI-prevention.md

```diff
@@ -0,0 +1,179 @@
+
+
+```ts
+public function getFilter($name)
+{
+        [snip]
+        foreach ($this->filterCallbacks as $callback) {
+        if (false !== $filter = call_user_func($callback, $name)) {
+            return $filter;
+        }
+    }
+    return false;
+}
+
+$output = $twig->render($_GET['name'],  array("first_name" => $user.first_name) );
+```
+
+#### Analysis
+
+In the above example, the user controls the template's content via the `name` GET parameter rather than a value passed into it.
+
+#### Impact
+
+An arbitrary shell command can be executed easily by registering `exec` as a filter callback and then calling `getFilter()`.
+
+An attacker can execute system commands using the "env" attribute of the "Twig_Environment" object.
+
+
+
+```ts
+Attacker’s input via name parameter,
+
+{{_self.env.registerUndefinedFilterCallback("exec")}}{{_self.env.getFilter("id")}}
+```
+
+#### Prevention
+
+Everything works fine if the name is passed into the template, as in the following example.
+
+```ts
+$output = $twig->render("Dear {name},", array("first_name" => $user.first_name) ); 
+```
+
+
+
+### Jinja2
+
+A vulnerability exists in the following Jinja2 code because the user-supplied variable `name` is concatenated to the template string. When an HTTP GET request is made, the page function accepts a `name` parameter and renders an HTML response with the contents of the name variable.
+
+
+
+```ts
+@app.route("/page")
+def page():
+    name = request.values.get('name')
+    output = Jinja2.from_string('Hello ' + name + '!').render()
+    return output
+```
+
+
+#### Analysis
+
+When an HTTP GET request is made, the page function accepts a 'name' parameter and renders an HTML response with the contents of the name variable.
+
+#### Impact
+
+As a result of this usage of templates, both XSS vulnerabilities and SSTI vulnerabilities are created. The following code block is an example of an attacker’s input via the `name` parameter.
+
+```ts
+{{config.__class__.__init__.__globals__['os'].popen('ls').read()}}
+```
+
+
+### Bypassing the blacklisting
+
+The problem with blacklists is that they can often be circumvented, which is terrible.
+ 
+```ts
+@app.route("/page")
+blacklist = ["__class__",”request[request.”]
+def page():
+    name = request.values.get('name')
+    for bad_string in blacklist:
+        if  bad_string in name:
+            return "HACKING ATTEMPT {}".format(bad_string), 400
+        else:
+            output = Jinja2.from_string('Hello ' + name + '!').render()
+            return output    
+```
+
+#### Analysis
+
+It is possible to retrieve the value of a new `cobalt` GET parameter using the `request.args.param`. An attacker can bypass `__class__` using this. Since we have also blacklisted `request.args.param`, we will use jinja2 native function called `| attr().` 
+
```

Comment:
```suggestion
A threat actor can bypass `__class__` and retrieve the value of a new GET parameter using `request.args.param`. Because we included `request.args.param` in the blacklist, we can use a native Jinja2 function `| attr().`
```